src/main/java/org/spongepowered/common/data/processor/value/ItemDyeColorValueProcessor.java

```diff
@@ -0,0 +1,79 @@
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.processor.value;
+
+import net.minecraft.item.EnumDyeColor;
+import net.minecraft.item.ItemStack;
+import org.spongepowered.api.data.DataTransactionResult;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.type.DyeColor;
+import org.spongepowered.api.data.type.DyeColors;
+import org.spongepowered.api.data.value.ValueContainer;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.common.data.processor.common.AbstractSpongeValueProcessor;
+import org.spongepowered.common.data.processor.data.DyeableDataProcessor;
+import org.spongepowered.common.data.value.SpongeValueFactory;
+import org.spongepowered.common.util.ColorUtil;
+
+import java.util.Optional;
+
+public class ItemDyeColorValueProcessor extends AbstractSpongeValueProcessor<ItemStack, DyeColor, Value<DyeColor>> {
+
+    public ItemDyeColorValueProcessor() {
+        super(ItemStack.class, Keys.DYE_COLOR);
+    }
+
+    @Override
+    protected boolean supports(ItemStack container) {
+        return DyeableDataProcessor.isDyeable(container.getItem());
+    }
+
+    @Override
+    protected Value<DyeColor> constructValue(DyeColor actualValue) {
+        return SpongeValueFactory.getInstance().createValue(Keys.DYE_COLOR, actualValue, DyeColors.BLACK);
+    }
+
+    @Override
+    protected boolean set(ItemStack container, DyeColor value) {
+        container.setItemDamage(((EnumDyeColor) (Object) value).getMetadata());
+        return true;
+    }
+
+    @Override
+    protected Optional<DyeColor> getVal(ItemStack container) {
+        return Optional.of((DyeColor) (Object) EnumDyeColor.byMetadata(container.getItemDamage()));
```

Comment:
Use `EnumDyeColor.byDyeDamage(stack.getMetadata())` - that's what `ItemDye` uses. This should fix what @Aaron1011 mentioned